src/node_port.py

```diff
@@ -0,0 +1,146 @@
+AuthMap = NamedTuple("AuthMap", protocol=AuthProtocol, mechanism=AuthMechanism)
+
+
+class NodePortManager:
+    """Manager for handling mongos Kubernetes resources for a single mongos pod."""
+
+    def __init__(
+        self,
+        charm: CharmBase,
+        pod_name: str,
+        namespace: str,
+    ):
+        self.charm = charm
+        self.pod_name = pod_name
+        self.app_name = "-".join(pod_name.split("-")[:-1])
+        self.namespace = namespace
+        self.nodeport_service_name = f"{self.app_name}-nodeport"
+        self.short_auth_mechanism_mapping: dict[AuthMechanism, str] = {
+            "SCRAM-SHA-512": "scram",
+            "OAUTHBEARER": "oauth",
+            "SSL": "ssl",
+        }
+
+    @cached_property
+    def client(self) -> Client:
+        """The Lightkube client."""
+        return Client(  # pyright: ignore[reportArgumentType]
+            field_manager=self.pod_name,
+            namespace=self.namespace,
+        )
+
+    # --- GETTERS ---
+
+    def get_pod(self, pod_name: str = "") -> Pod:
+        """Gets the Pod via the K8s API."""
+        # Allows us to get pods from other peer units
+        pod_name = pod_name or self.pod_name
+
+        return self.client.get(
+            res=Pod,
+            name=self.pod_name,
+        )
+
+    def get_service(self, service_name: str) -> Service | None:
+        """Gets the Service via the K8s API."""
+        return self.client.get(
+            res=Service,
+            name=service_name,
+        )
+
+    def get_node_port(
+        self,
+        service: Service,
+        auth_map: AuthMap,
+    ) -> int:
+        """Gets the NodePort number for the service via the K8s API."""
+        if not service.spec or not service.spec.ports:
+            raise Exception("Could not find Service spec or ports")
+
+        for port in service.spec.ports:
+            if (
+                auth_map.protocol.lower().replace("_", "-") in port.name
+                and self.short_auth_mechanism_mapping[auth_map.mechanism] in port.name
+            ):
+                return port.nodePort
+
+        raise Exception(
+            f"Unable to find NodePort using {auth_map.protocol} and {auth_map.mechanism} for the {service} service"
+        )
+
+    def build_node_port_services(self, port: str) -> Service:
+        """Builds a ClusterIP service for initial client connection."""
+        pod = self.get_pod(pod_name=self.pod_name)
+        if not pod.metadata:
+            raise Exception(f"Could not find metadata for {pod}")
+
+        return Service(
+            metadata=ObjectMeta(
+                name=self.nodeport_service_name,
+                namespace=self.namespace,
+                # owned by the StatefulSet
+                ownerReferences=pod.metadata.ownerReferences,
+            ),
+            spec=ServiceSpec(
+                type="NodePort",
+                selector={"app.kubernetes.io/name": self.app_name},
```

Comment:
**todo:** Based on my above comment and limited understanding, I **think** you want to change this selector to be the pod, not the unit.

```python
selector={"statefulset.kubernetes.io/pod-name": self.pod_name}
```